### PR TITLE
network: move remaining socket syscalls to socket

### PR DESCRIPTION
**Pull request by florincoras**
- add functions to set/get socket options directly on socket.
- add function to update socket blocking flag

Signed-off-by: Florin Coras <fcoras@cisco.com>

Risk Level: Medium
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/network/socket_impl.h [hunk 1]
```diff
@@ -48,6 +48,11 @@ Address::InstanceConstSharedPtr addressFromFd(os_fd_t fd);
  */
 Address::InstanceConstSharedPtr peerAddressFromFd(os_fd_t fd);
 
+/**
+ * Retrieve host name (@see man 2 gethostname)
+ */
+Api::SysCallIntResult getHostName(char* name, size_t length);
```
[the remark this answers is not in the document]
Comment (florincoras) [reply source/common/network/socket_impl.h:54]: You're definitely right. Not sure what I was thinking ... Reverting change.

File: include/envoy/network/socket.h [hunk 2]
```diff
@@ -113,6 +118,23 @@ class Socket {
...
+  /**
+   * Retrieves option from underlying socket (@see man 2 getsockopt)
+   */
+  virtual Api::SysCallIntResult getSocketOption(int level, int optname, void* optval,
+                                                socklen_t* optlen) PURE;
+
+  /**
+   * Toggle socket blocking state
+   */
+  virtual Api::SysCallIntResult setBlocking(bool blocking) PURE;
```
[the remark this answers is not in the document]
Comment (florincoras) [reply include/envoy/network/socket.h:136]: Done!

**Comment by ggreenway** [comment]
@florincoras Can you look into the CI failure?

/wait

**Comment by florincoras** [comment]
> @florincoras Can you look into the CI failure?
> 
Trying to figure this out but it seems that this only fails with asan/tsan. Locally, just running the test(s) also passes. Is there any way to run asan locally?

**Pull request merged**